login/service.go

```diff
@@ -152,6 +132,62 @@ func (gh *gitHubOAuth) Perform(ctx *app.AuthorizeLoginContext) error {
 	return ctx.TemporaryRedirect()
 }
 
+type identityError struct {
+	message interface{}
+	code    int
+}
+
+const (
+	noPrimaryEmail            int = 1
+	associatedUserNotFound    int = 2
+	unauthorizedGitHubRequest int = 3
+)
+
+func (gh *gitHubOAuth) getUserIdentity(ctx context.Context, ghtoken *oauth2.Token) (*account.Identity, *identityError) {
+	emails, err := gh.getUserEmails(ctx, ghtoken)
+	fmt.Println(emails)
+
+	primaryEmail := filterPrimaryEmail(emails)
+	if primaryEmail == "" {
+		fmt.Println("No primary email found?! ", emails)
+		return nil, &identityError{code: noPrimaryEmail, message: PrimaryEmailNotFoundError}
+	}
+	users, err := gh.users.Query(account.UserByEmails([]string{primaryEmail}), account.UserWithIdentity())
+	if err != nil {
+		return nil, &identityError{code: associatedUserNotFound, message: AssociatedUserNotFound + " " + err.Error()}
+	}
+	ghUser, err := gh.getUser(ctx, ghtoken)
+	if err != nil {
+		fmt.Println(err)
+		jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrUnauthorized(err.Error()))
+		return nil, &identityError{code: unauthorizedGitHubRequest, message: jerrors}
+	}
+	fmt.Println(ghUser)
+	var identity account.Identity
+	if len(users) == 0 {
+		// No User found, create new User and Identity
+		identity = createIdentity(*ghUser)
+		gh.identities.Create(ctx, &identity)
+		gh.users.Create(ctx, &account.User{Email: primaryEmail, Identity: identity})
+	} else {
+		identity = users[0].Identity
+		// let's update the current identity with the fullname and avatar from GitHub,
+		// in case the user changed them since the last time he logged in here
+		name := ghUser.Name
+		if name == "" {
+			identity.FullName = ghUser.Login
+		} else {
+			identity.FullName = ghUser.Name
+		}
+		identity.ImageURL = ghUser.AvatarURL
+		identity.UpdatedAt = time.Now()
+		gh.identities.Save(ctx, &identity)
+	}
+
+	fmt.Println("Identity: ", identity)
```

Comment:
If you really need to print something, use `log.Println(....)` but I doubt that you even want to do that here.